parser/lex_test.go

```diff
@@ -0,0 +1,266 @@
+	{"gte", ">=", []item{item{itemGTE, 0, ">="}, tEOF}},
+
+	{"not", "!", []item{item{itemNot, 0, "!"}, tEOF}},
+	{"not equals", "!=", []item{item{itemNotEquals, 0, "!="}, tEOF}},
+
+	{"dot access", ".", []item{item{itemDotAccessOperator, 0, "."}, tEOF}},
+	{"arrow access", "->", []item{item{itemArrowAccessOperator, 0, "->"}, tEOF}},
+
+	{"null or access", "??", []item{item{itemNullOrValueOperator, 0, "??"}, tEOF}},
+	{"null dot access", "?.", []item{item{itemNullDotAccessOperator, 0, "?."}, tEOF}},
+
+	{"true literal", "true", []item{tTrue, tEOF}},
+	{"false literal", "false", []item{tFalse, tEOF}},
+
+	{"this keyword", "this", []item{item{itemKeyword, 0, "this"}, tEOF}},
+	{"class keyword", "class", []item{item{itemKeyword, 0, "class"}, tEOF}},
+	{"interface keyword", "interface", []item{item{itemKeyword, 0, "interface"}, tEOF}},
+
+	{"not class keyword", "classy", []item{item{itemIdentifer, 0, "classy"}, tEOF}},
+
+	{"string literal", "'string test'", []item{item{itemStringLiteral, 0, "'string test'"}, tEOF}},
+	{"string literal", "\"string test\"", []item{item{itemStringLiteral, 0, "\"string test\""}, tEOF}},
+
+	{"string literal with escape", "\"string \\\" test\"", []item{item{itemStringLiteral, 0, "\"string \\\" test\""}, tEOF}},
+
+	{"string literal with newline", "\"string\ntest\"", []item{item{itemError, 0, "Unterminated string literal"}}},
+
+	{"template string literal", "`string\ntest`", []item{item{itemTemplateStringLiteral, 0, "`string\ntest`"}, tEOF}},
+
+	{"identifier", "someident", []item{item{itemIdentifer, 0, "someident"}, tEOF}},
+	{"identifier 2", "someIdent", []item{item{itemIdentifer, 0, "someIdent"}, tEOF}},
+	{"identifier 3", "SomeIdent", []item{item{itemIdentifer, 0, "SomeIdent"}, tEOF}},
+	{"identifier 4", "SOMEIDENT", []item{item{itemIdentifer, 0, "SOMEIDENT"}, tEOF}},
+	{"identifier 5", "SomeIdent1235", []item{item{itemIdentifer, 0, "SomeIdent1235"}, tEOF}},
+	{"identifier 6", "someIdent_1235", []item{item{itemIdentifer, 0, "someIdent_1235"}, tEOF}},
+	{"identifier 7", "_someIdent_1235", []item{item{itemIdentifer, 0, "_someIdent_1235"}, tEOF}},
+
+	{"unicode identifier", "םש", []item{item{itemIdentifer, 0, "םש"}, tEOF}},
+
+	{"numeric literal", "123", []item{item{itemNumericLiteral, 0, "123"}, tEOF}},
+	{"numeric literal 2", "-123", []item{item{itemNumericLiteral, 0, "-123"}, tEOF}},
+	{"numeric literal 3", "123.56", []item{item{itemNumericLiteral, 0, "123.56"}, tEOF}},
+	{"numeric literal 4", "-123.56", []item{item{itemNumericLiteral, 0, "-123.56"}, tEOF}},
+	{"numeric literal 5", "123e56", []item{item{itemNumericLiteral, 0, "123e56"}, tEOF}},
+	{"numeric literal 6", "123e-56", []item{item{itemNumericLiteral, 0, "123e-56"}, tEOF}},
+	{"numeric literal 7", "0xfe", []item{item{itemNumericLiteral, 0, "0xfe"}, tEOF}},
+
+	// Complex tests.
+	{"dot expression test", "this.foo", []item{
+		item{itemKeyword, 0, "this"},
+		item{itemDotAccessOperator, 0, "."},
+		item{itemIdentifer, 0, "foo"},
+		tEOF}},
```

Comment:
The indentation on this feels weird because hanging indentation isn't a usual sight in Go. I think I'd do it like this, but it'd add _many_ lines.

```go
	{
		"dot expression test",
		"this.foo",
		[]item{
			item{itemKeyword, 0, "this"},
			item{itemDotAccessOperator, 0, "."},
			item{itemIdentifer, 0, "foo"},
			tEOF,
		},
	},
```